feat: Letterhead Scripts & fix disappearing header in pdf

marination opened the pull request:
`no-docs`

## Feature
- Allow script injection into header/footer.html to allow manipulation of styles using page numbers/args received by wkhtmltopdf (because these args are only received/substituted in header and footer html files generated by wkhtmltopdf)
- Eg: we want the page numbers to be red but only on the first page
- Add the following Letterhead HTML
  ```html
  <div id="header-html">
	    <div class="header-content">
	        Page <span class="page"></span> of <span class="topage"></span>
	    </div>
  </div>
   ```
- Add some valid JS in **Header Script** / **Footer Script**. The idea is to have the header be in red on the first page in a pdf and in blue via print to make sure the script is visibly applied
  ```js
	var vars = {};
	var query_strings_from_url = document.location.search.substring(1).split('&');
	for (var query_string in query_strings_from_url) {
		if (query_strings_from_url.hasOwnProperty(query_string)) {
			var temp_var = query_strings_from_url[query_string].split('=', 2);
			vars[temp_var[0]] = decodeURI(temp_var[1]);
		}
	}
	var el = document.getElementsByClassName("header-content");
	
	if (vars["page"] == "1") {
		// "page" will be available only in pdf via wkhtmltopdf
		el[0].style.color = "red";
	} else {
	    el[0].style.color = "blue";
	}
  ```

**Result (Note the header):**
- PDF
   <img width="600" alt="Screenshot 2023-12-21 at 5 31 41 PM" src="https://github.com/frappe/frappe/assets/25857446/3c9702d7-3f44-4046-9bab-42127ba8e840">
- Print (does not have page numbers substituted in custom header because it is not rendered via wkhtmltopdf, treat this as an example)
  <img width="472" alt="Screenshot 2023-12-21 at 5 32 22 PM" src="https://github.com/frappe/frappe/assets/25857446/bb3bc6f2-f244-4f88-b090-b02566ee6f83">

## Issue:
- Create a **Letterhead** record that attempts to show the Page number (as per the `Footers And Headers` section in https://wkhtmltopdf.org/usage/wkhtmltopdf.txt)
   <img width="607" alt="Screenshot 2023-11-20 at 3 22 05 PM" src="https://github.com/frappe/frappe/assets/25857446/d20eb29e-b90f-4fcd-91de-5e4a83703dcf">
- Letter has `id` as `header-html` (extracted in `pdf.py`) so that it is rendered as a separate `header.html` by wkhtmltopdf and so that it receives the "page", "topage" variables for substitution
- Use this letterhead while printing with any format (Standard format taken in eg.). 
-  **The Letterhead does not show up**
   ![2023-11-20 14 20 09](https://github.com/frappe/frappe/assets/25857446/a0b023ec-f3eb-41a9-83ea-041534ecf0ee)
- **Source:** `header-html` is extracted from the html print document to avoid duplicates, but it is done before wkhtmltopdf gets a chance to render the `header-html` section into a `header.html` file


## Fix:
- Make sure 'header-html' is extracted from document after it is rendered as html and loaded in wkhtmltopdf options, else it goes missing
  ![2023-11-20 14 21 47](https://github.com/frappe/frappe/assets/25857446/7baaa9bb-4b57-4061-ad3e-12d475a766c2)
- Use `if` conditions to secure js script. Errors cause silent breakage and hence the scripts are not effective/not run. There is no feedback

marination committed:
feat: Dynamic headers/footers & fix disappearing header

- fix: Make sure 'header-html' is extracted from document after it is rendered as html and loaded in wkhtmltopdf options, else it goes missing
- feat: Allow script injection into header/footer.html to allow manipulation of styles using page numbers/args received by wkhtmltopdf

marination commented:
Will open a separate PR for develop due its code structure being different

marination committed:
feat: Include letterhead scripts in print (html)

- misc: also validate letterhead scripts

marination committed:
chore: Add helper instructions & re-arrange fields

siddhpant commented:
Does this also fix #23090?

marination commented:
@siddhpant I cant reproduce https://github.com/frappe/frappe/issues/23090. Image letterhead seems to work fine via a custom print format. Perhaps there is an issue in the custom print format?

siddhpant commented:
I'll add photos in that issue.

Though you might be correct, it does seem to be a part of bigger issues in the custom print format codebase, as there are many issues...

PatrickDEissler commented:
Comment against inactivity...

ankush commented:
@marination can you please keep v15/develop port ready.

ankush commented on frappe/utils/pdf.py:
```diff
@@ -23,9 +23,9 @@
-	html, options = prepare_options(html, options)
+	html, options = prepare_options(html, options, letterhead=letterhead)
 
 	options.update({"disable-javascript": "", "disable-local-file-access": ""})
```
Do scripts even work because of this? 

This is disabled for security reasons https://github.com/frappe/frappe/commit/36bc812c9493f578a3a39f059252e6c967d5decc

marination replied:
From the proof I added in the description, yup. Maybe the flag does not apply to headers and footers?
Because there is pre-existing js in `pdf_header_footer.html` that is used for page substitution as per https://wkhtmltopdf.org/usage/wkhtmltopdf.txt (see "Footers and Headers"). I extended the script a bit more.

ankush replied:
@maharshivpatel can you confirm this please.

maharshivpatel replied:
yup, JS works in header/footer even after disabling it.

i guess this is intentional because page substitution in header/footer uses JS.

marination commented on frappe/www/printview.html:
```diff
@@ -31,14 +31,17 @@
-			footer_html.classList.remove('visible-pdf');
-			footer_html.style.order = 1;
-			footer_html.style.marginTop = '20px';
+			if (footer_html)  {
```
```suggestion
			const footer_html = document.getElementById('footer-html');
			if (footer_html)  {
```

marination committed:
fix: Missing `footer_html ` declaration

marination commented:
@maharshivpatel bump

barredterra commented on frappe/printing/doctype/letter_head/letter_head.js:
```diff
@@ -2,7 +2,61 @@
+	el[0].textContent += " : " + vars["date"];
+}
+</code>
+</pre>`;
```
```suggestion
const INSTRUCTIONS = `<h4>${__("Letter Head Scripts")}</h4>
<p>${ __("Header/Footer scripts can be used to add dynamic behaviours.") }</p>

<pre>
<code>
// ${ __("The following Header Script will add the current date to an element in 'Header HTML' with class 'header-content'") }
var el = document.getElementsByClassName("header-content");
if (el.length > 0) {
	el[0].textContent += " " + new Date().toGMTString();
}
</code>
</pre>

<p>${ __("You can also access wkhtmltopdf variables (valid only in PDF print):") }</p>

<pre>
<code>
// ${ __("Get Header and Footer wkhtmltopdf variables") }
// ${ __("Snippet and more variables: {0}", ["https://wkhtmltopdf.org/usage/wkhtmltopdf.txt"]) }
var vars = {};
var query_strings_from_url = document.location.search.substring(1).split('&');
for (var query_string in query_strings_from_url) {
	if (query_strings_from_url.hasOwnProperty(query_string)) {
		var temp_var = query_strings_from_url[query_string].split('=', 2);
		vars[temp_var[0]] = decodeURI(temp_var[1]);
	}
}

var el = document.getElementsByClassName("header-content");
if (el.length > 0 && vars["page"] == 1) {
	el[0].textContent += " : " + vars["date"];
}
</code>
</pre>`;
```

barredterra commented:
The header script seems to get executed twice:

![Bildschirmfoto 2024-01-05 um 13 19 59](https://github.com/frappe/frappe/assets/14891507/70262d6d-827a-4776-9e0b-82e903716943)
![Bildschirmfoto 2024-01-05 um 13 20 28](https://github.com/frappe/frappe/assets/14891507/061c8477-e075-434d-8c94-87a0e39f33e6)

barredterra commented:
I was able to reproduce the example from the PR description (coloring page numbers).

PatrickDEissler commented:
PR is still relevant.

marination committed:
fix: Translate Letterhead information string

marination committed:
fix: Custom Script runs twice

- `frappe.get_print` calls `get_response_content("printview")` which calls `www/printview` that adds header and footer scripts
- Now it gets added again via `pdf_header_footer.html` which causes the script to run twice
- Header and Footer html are also added twice but then removed and handled in pdf.py (because printview + wkhtml processing)
- Misc: Use walrus operator
- Remove passed `letterhead` to insert script via pdf.py, printview handles it

marination commented:
> The header script seems to get executed twice:

@barredterra this is fixed

marination committed:
fix: Extract header/footer html into `content` to simplify things

- Extract header/footer html into `content` first, then remove any instances from the main content since it is already rendered via the extracted header/footer html

marination committed:
fix: Check if header/footer html exists before trying to extract

- issue: AttributeError: 'NoneType' object has no attribute 'extract', when no letterhead html was present
